File: ドメイン駆動/C/Value_Object.c
/*

■Value Object（値オブジェクト）

Cではイミュータブルは保証できないが、readonly的な使い方は可能。

*/

typedef struct {
    char email[100];

} Email;

int is_valid_email(const Email* email) {
    return strchr(email->email, '@') != NULL;
}

File: ドメイン駆動/PHP/Value_Object.php
<?php

/*
識別子を持たず、属性の集合であり、イミュータブル（不変）なオブジェクト。
例：金額、住所、日付など。
*/

final class Email
{
    public function __construct(private readonly string $value)
    {
        if(!filter_var($value, FILTER_VALIDATE_EMAIL)) {
            throw new InvalidArgumentException("Invalid email: $value");
        }
    }

    public function getValue(): string
    {
        return $this->value;
    }
}



?>